### PR TITLE
FIP-78a: Metagovernance - ANGLE

### PR DESCRIPTION
**Pull request by eswak**
The purpose of this PR is to enable meta-governance with the protocol-owned governance tokens :
- ANGLE
- INDEX (already live)

# SnapshotDelegatorPCVDeposit
This contract existed already, I just moved it to a new `metagov` folder. The constructor also now include the setting of a contract admin role `_setContractAdminRole(keccak256("METAGOV_ADMIN_ROLE"));`, to enable OA to change delegates.

Used for `INDEX` (already deployed on-chain)

# AngleDelegatorPCVDeposit
This contract is a `SnapshotDelegatorPCVDeposit` that can also vote-escrow tokens (`VoteEscrowTokenManager`), vote and propose on an OZ governor (`OZGovernorVoter`), and stake tokens in gauges and claim their rewards (`LiquidityGaugeManager`).

Used for `ANGLE`.

### CODE REVIEW

File: contracts/metagov/SnapshotDelegatorPCVDeposit.sol [hunk 1]
```diff
@@ -42,6 +42,9 @@ contract SnapshotDelegatorPCVDeposit is PCVDeposit {
         token = _token;
         spaceId = _spaceId;
         _delegate(_initialDelegate);
+
+        // Use a shared role to manage meta-governance related actions
+        _setContractAdminRole(keccak256("METAGOV_ADMIN_ROLE"));
```
Comment (Joeysantoro) [inline contracts/metagov/SnapshotDelegatorPCVDeposit.sol:47]: Lets start using TribeRoles.sol and the onlyRole modifier instead

File: contracts/metagov/utils/LiquidityGaugeManager.sol [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    /// @notice Stake tokens in a gauge
+    /// @param gaugeAddress the address of a gauge to stake tokens in
+    /// @param token the address of the token to stake in the gauge
+    /// @param amount the amount of tokens to stake in the gauge
+    function stakeInGauge(
+        address gaugeAddress,
+        address token,
+        uint256 amount
+    ) public whenNotPaused onlyPCVController {
```
Comment (Joeysantoro) [inline contracts/metagov/utils/LiquidityGaugeManager.sol:43]: Lets make a GAUGE_ADMIN role?

File: contracts/metagov/utils/OZGovernorVoter.sol [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    ) external onlyGovernorOrAdmin {
+        governor.propose(targets, values, calldatas, description);
+    }
+
+    /// @notice cast a vote on a given proposal on the target governor.
+    function castVote(
+        IOZGovernor governor,
+        uint256 proposalId,
+        uint8 support
+    ) external onlyGovernorOrAdmin {
```
Comment (Joeysantoro) [inline contracts/metagov/utils/OZGovernorVoter.sol:36]: these can be METAGOVERNANCE_ADMIN only

File: contracts/metagov/utils/VoteEscrowTokenManager.sol [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        liquidToken = _liquidToken;
+        veToken = _veToken;
+        MAXTIME = _maxTime;
+    }
+
+    /// @notice Deposit tokens to get veTokens. Set lock duration to MAXTIME.
+    /// @dev permissionless: anyone can lock tokens and extend the lock
+    /// duration at anytime. The only way to withdraw tokens will be to
+    /// pause this contract for MAXTIME and then call exitLock().
+    function lock() external whenNotPaused {
```
Comment (Joeysantoro) [inline contracts/metagov/utils/VoteEscrowTokenManager.sol:49]: should be behind a low-clearance VOTE_ESCROW role imo

Comment (eswak) [reply contracts/metagov/utils/VoteEscrowTokenManager.sol:49]: I'll make a `METAGOVERNANCE_TOKEN_STAKING` role, that we'll use for AAVE/stkAAVE and CVX/vlCVX too. The mechanisms are very similar.

File: contracts/metagov/AngleDelegatorPCVDeposit.sol [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        address _initialDelegate
+    ) SnapshotDelegatorPCVDeposit(
+        _core,
+        IERC20(0x31429d1856aD1377A8A0079410B297e1a9e214c2), // ANGLE
+        keccak256("anglegovernance.eth"),
+        _initialDelegate
+    ) VoteEscrowTokenManager(
+        IERC20(0x31429d1856aD1377A8A0079410B297e1a9e214c2), // ANGLE
+        IVeToken(0x0C462Dbb9EC8cD1630f1728B2CFD2769d09f0dd5), // veANGLE
+        4 * 365 * 86400 // 4 years
```
Comment (ElliotFriedman) [inline contracts/metagov/AngleDelegatorPCVDeposit.sol:27]: You can just use solidity keyword `day` instead of putting 86,400

**Comment by ElliotFriedman** [comment]
It would be great to see unit tests on all this new solidity

File: contracts/metagov/AngleDelegatorPCVDeposit.sol (hunk 5 above)
Comment (eswak) [reply contracts/metagov/AngleDelegatorPCVDeposit.sol:27]: I'd like to keep the same exact wording as in the veANGLE token's code if you don't mind

**Comment by eswak** [comment]
> It would be great to see unit tests on all this new solidity

I added a lot of e2e tests for the various new contracts, they interact with real on-chain governors, gauges, and veTokens, to validate the behavior

File: proposals/dao/fip_78a.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  await gaugeLensAgEurUniswapGauge.deployTransaction.wait();
+  logging && console.log('gaugeLensAgEurUniswapGauge:', gaugeLensAgEurUniswapGauge.address);
+
+  const uniswapLensFactory = await ethers.getContractFactory('UniswapLens');
+  const uniswapLensAgEurUniswapGauge = await uniswapLensFactory.deploy(
+    gaugeLensAgEurUniswapGauge.address, // deposit to read from
+    addresses.core, // core
+    addresses.chainlinkEurUsdOracleWrapper, // oracle
+    ethers.constants.AddressZero, // no backup oracle
+    false // fei is token1 in the pool
```
Comment (thomas-waite) [inline proposals/dao/fip_78a.ts:53]: With previous comment above, I'd try to push this `_isFeiToken0` bool down and have the contract programmatically figure it out

File: test/integration/setup/index.ts [hunk 7]
```diff
@@ -147,6 +150,52 @@ export class TestEndtoEndCoordinator implements TestCoordinator {
       await proposal.simulate();
     }
 
+    if (config.category === ProposalCategory.DEBUG) {
```
Comment (thomas-waite) [inline test/integration/setup/index.ts:153]: Perhaps not for this PR, but this `applyUpgrade` function is getting unwieldy and hard to follow. Think we refactor to use `switch` statements and encapsulate each `if (config.category === ProposalCategory.X)` into it's own function

File: test/integration/setup/index.ts [hunk 8]
```diff
@@ -147,6 +150,52 @@ export class TestEndtoEndCoordinator implements TestCoordinator {
...
+      console.log('Simulating DAO proposal in DEBUG mode (step by step)...');
+      console.log('  Title: ', config.proposal.title);
+
+      const signer = await getImpersonatedSigner(contracts.feiDAOTimelock.address);
+      await forceEth(contracts.feiDAOTimelock.address);
+
+      let totalGasUsed = 0;
+      for (let i = 0; i < config.proposal.commands.length; i++) {
```
Comment (thomas-waite) [inline test/integration/setup/index.ts:161]: I find this logic in lines 161 - 194 very hard to follow. 

It's some form of parser for a DAO proposal? I would encapsulate into it's own function with a description of what it does and perhaps include an example input and output. 

If it's significantly extending e2e test functionality, I'd probably break out into a seperate PR

File: test/integration/setup/index.ts [hunk 9]
```diff
@@ -147,6 +150,52 @@ export class TestEndtoEndCoordinator implements TestCoordinator {
...
+        const to = contractAddresses[cmd.target] || cmd.target;
+        const value = cmd.values;
+        console.log('    Target:', cmd.target, '[' + to + ']');
+        console.log('    Method:', cmd.method);
+        types.forEach((type, i) => {
+          console.log('      Argument', i, '[' + type + ']', cmdArguments[i]);
+        });
+        //console.log('    Value:', value);
+        //console.log('    Calldata:', calldata);
+        process.stdout.write('    Calling... ');
```
Comment (thomas-waite) [inline test/integration/setup/index.ts:189]: Why use `process.stdout` and not just `console.log`? There's also commented out `console.log`s

File: test/integration/tests/metagov.ts [hunk 10]
```diff
@@ -0,0 +1,457 @@
...
+      logging,
+      deployAddress,
+      version: 1
+    };
+    e2eCoord = new TestEndtoEndCoordinator(config, proposals);
+    ({ contracts } = await e2eCoord.loadEnvironment());
+  });
+
+  describe('utils/VoteEscrowTokenManager.sol', function () {
+    let locker: any;
```
Comment (thomas-waite) [inline test/integration/tests/metagov.ts:38]: `locker` should have a specific contract type rather than `any`

File: test/integration/tests/metagov.ts [hunk 11]
```diff
@@ -0,0 +1,457 @@
...
+        const holder = '0x4f91F01cE8ec07c9B1f6a82c18811848254917Ab';
+        await forceEth(holder);
+        const signer = await getImpersonatedSigner(holder);
+        await contracts.angle.connect(signer).transfer(locker.address, e18(1_000_000));
+        await locker.lock();
+        expect(await contracts.angle.balanceOf(locker.address)).to.be.equal('0');
+        expect(await contracts.veAngle.balanceOf(locker.address)).to.be.at.least(e18(10_900_000));
+      });
+      it('fast-forward 4 years', async function () {
+        // fast-forwarding 4 years directly causes a revert without reason,
```
Comment (thomas-waite) [inline test/integration/tests/metagov.ts:125]: Hmm, do we know why it causes a revert?

File: test/integration/tests/metagov.ts [hunk 12]
```diff
@@ -0,0 +1,457 @@
...
+        expect(await contracts.angle.balanceOf(locker.address)).to.be.equal('0');
+        expect(await contracts.veAngle.balanceOf(locker.address)).to.be.equal('0');
+        await locker.exitLock();
+        expect(await contracts.angle.balanceOf(locker.address)).to.be.at.least(e18(10_900_000));
+        expect(await contracts.veAngle.balanceOf(locker.address)).to.be.equal('0');
+      });
+    });
+  });
+
+  describe('utils/OZGovernorVoter.sol', function () {
```
Comment (thomas-waite) [inline test/integration/tests/metagov.ts:147]: I don't like nesting multiple different test blocks into the same file. Why not create a parent `metagov` directory and then put each of these `describe` blocks into their own file? It would split things up more + reduce the amount of code to review in each file.

If some `before` block code is needed across each `describe` then we can break it out into a `fixture` and import

**Review by thomas-waite — COMMENTED** [review]
Main comments:
- Question about whether we are effectively going to be continuously locking tokens with this implementation
- Would break out the parser/debug mode you've added into the e2e tests to a seperate PR. Will reduce this PR size + make it easier to go through the parser/debug mode. Not sure adding a new `ProposalCategory` of `DEBUG` is the optimal way to do it
- Would break out the `metagov.ts` tests into several different test files with a parent `metagov` directory. Will reduce the file size and make it easier to review


On a seperate note, I know the e2e tests for this work have been taking a long time to run and made development painful. In the `Backup Oracles` PR, I've added forge based integration tests for development. They run significantly faster than our `TypeScript` e2e tests, and will likely help develop like this

File: contracts/metagov/utils/VoteEscrowTokenManager.sol [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        MAXTIME = _maxTime;
+    }
+
+    /// @notice Deposit tokens to get veTokens. Set lock duration to MAXTIME.
+    /// The only way to withdraw tokens will be to pause this contract
+    /// for MAXTIME and then call exitLock().
+    function lock() external whenNotPaused onlyTribeRole(TribeRoles.METAGOVERNANCE_TOKEN_STAKING) {
+        uint256 tokenBalance = liquidToken.balanceOf(address(this));
+        uint256 locked = veToken.locked(address(this));
+        uint256 lockHorizon = ((block.timestamp + MAXTIME) / WEEK) * WEEK;
```
[the remark this answers is not in the document]
Comment (eswak) [reply contracts/metagov/utils/VoteEscrowTokenManager.sol:56]: the lock horizon is rounded down to entire weeks in the veToken contract, I am doing the same here to make sure I can test that the lock horizon is > what is already locked, because otherwise the call to increase_lock_time reverts (so I don't want to call it, if it's to set the same lock horizon)

File: contracts/metagov/utils/VoteEscrowTokenManager.sol [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+        IERC20 _liquidToken,
+        IVeToken _veToken,
+        uint256 _maxTime
+    ) {
+        liquidToken = _liquidToken;
+        veToken = _veToken;
+        MAXTIME = _maxTime;
+    }
+
+    /// @notice Deposit tokens to get veTokens. Set lock duration to MAXTIME.
```
[the remark this answers is not in the document]
Comment (eswak) [reply contracts/metagov/utils/VoteEscrowTokenManager.sol:50]: that's a good idea, I'll make the lock duration configurable (and name it differently, MAXTIME refers to the veToken's MAXTIME, which is the maximum amount of seconds a token can be vote-escrowed)

File: contracts/metagov/utils/VoteEscrowTokenManager.sol [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+        // No additional tokens to lock, just refresh duration to MAXTIME
+        else if (tokenBalance == 0 && locked != 0) {
+            veToken.increase_unlock_time(lockHorizon);
+        }
+        // If tokenBalance == 0 and locked == 0, there is nothing to do.
+    }
+
+    /// @notice Exit the veToken lock. For this function to be called and not
+    /// revert, tokens had to be locked previously, and the contract must have
+    /// been paused for MAXTIME in order to prevent lock extensions
```
[the remark this answers is not in the document]
Comment (eswak) [reply contracts/metagov/utils/VoteEscrowTokenManager.sol:80]: yes, that's how veTokens are intended to work 😓 also your voting power decreases linearly over time, so if you don't renew your 4-year lock, after 2 years you only have half of your voting power left. Splitting your locks in different tranches is not desirable because you'd need multiple addresses & the tranches that expire soon don't provide voting power anymore

File: contracts/pcv/angle/GaugeLens.sol [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+        gaugeAddress = _gaugeAddress;
+        stakerAddress = _stakerAddress;
+        balanceReportedIn = ILiquidityGauge(_gaugeAddress).staking_token();
+    }
+
+    function balance() public view override returns(uint256) {
+        return ILiquidityGauge(gaugeAddress).balanceOf(stakerAddress);
+    }
+
+    function resistantBalanceAndFei() public view override returns(uint256, uint256) {
```
[the remark this answers is not in the document]
Comment (eswak) [reply contracts/pcv/angle/GaugeLens.sol:34]: "resistant" balance means that it can't be manipulated with a flashloan or something transitory like that

I'll add natspec

The lens reports the amount of tokens staked in a gauge, and reports in this token. Unless the token staked in the gauge is FEI, the FEI balance will always be zero. And there are no gauges where naked FEI is staked, it's only LP tokens & stuff like that. The lens simply reports the number of LP tokens in that case. And there is a second "uniswap lens" that, given a balance of LP tokens, deduces a number of tokens & fei held within them.

File: contracts/pcv/uniswap/UniswapLens.sol [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+
+    /// @notice true if FEI is token0 in the Uniswap pool
+    bool public immutable feiIsToken0;
+
+    constructor(
+        address _depositAddress,
+        address _core,
+        address _oracle,
+        address _backupOracle,
+        bool _feiIsToken0
```
[the remark this answers is not in the document]
Comment (eswak) [reply contracts/pcv/uniswap/UniswapLens.sol:29]: good idea, I copied the logic from the BPT lens, but I like your improvement suggestion

File: test/integration/tests/metagov.ts (hunk 10 above)
Comment (eswak) [reply test/integration/tests/metagov.ts:38]: I'll type

File: test/integration/tests/metagov.ts (hunk 11 above)
Comment (eswak) [reply test/integration/tests/metagov.ts:136]: I think it has to do with the veToken implementation, some mathematical formulas does not work properly if the checkpoints are too spaced in time, or something like that. It reverts inside the checkpoint() call of the veToken contract, not on our side

File: test/integration/tests/metagov.ts (hunk 12 above)
Comment (eswak) [reply test/integration/tests/metagov.ts:158]: sounds good, I'll split the files

File: contracts/core/TribeRoles.sol [hunk 18]
```diff
@@ -58,6 +58,21 @@ library TribeRoles {
...
+    /// @notice allows to manage locking of vote-escrowed tokens, and staking/unstaking
+    /// governance tokens from a pre-defined contract in order to eventually allow voting.
+    /// Examples: ANGLE <> veANGLE, AAVE <> stkAAVE, CVX <> vlCVX, CRV > cvxCRV.
+    bytes32 internal constant METAGOVERNANCE_TOKEN_STAKING = keccak256("METAGOVERNANCE_TOKEN_STAKING");
+
+    /// @notice manages whitelisting of gauges where the protocol's tokens can be staked
+    bytes32 internal constant METAGOVERNANCE_GAUGE_ADMIN = keccak256("METAGOVERNANCE_GAUGE_ADMIN");
+
+    /// @notice manages staking to & unstaking from gauges of the protocol's tokens
+    bytes32 internal constant METAGOVERNANCE_GAUGE_STAKING = keccak256("METAGOVERNANCE_GAUGE_STAKING");
```
Comment (Joeysantoro) [inline contracts/core/TribeRoles.sol:74]: There is some redundancy here, can we keep this to two roles?

File: contracts/metagov/AngleDelegatorPCVDeposit.sol [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+pragma solidity ^0.8.0;
+
+import "./SnapshotDelegatorPCVDeposit.sol";
+import "./utils/VoteEscrowTokenManager.sol";
+import "./utils/LiquidityGaugeManager.sol";
+import "./utils/OZGovernorVoter.sol";
+
+/// @title ANGLE Token PCV Deposit
+/// @author Fei Protocol
+contract AngleDelegatorPCVDeposit is SnapshotDelegatorPCVDeposit, VoteEscrowTokenManager, LiquidityGaugeManager, OZGovernorVoter {
```
Comment (Joeysantoro) [inline contracts/metagov/AngleDelegatorPCVDeposit.sol:11]: does this still need OZGovernorVoter?

File: contracts/metagov/AngleDelegatorPCVDeposit.sol [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+        _initialDelegate
+    ) VoteEscrowTokenManager(
+        IERC20(ANGLE_TOKEN), // liquid token
+        IVeToken(ANGLE_VE_TOKEN), // vote-escrowed token
+        4 * 365 * 86400 // vote-escrow time = 4 years
+    ) LiquidityGaugeManager(ANGLE_GAUGE_MANAGER) OZGovernorVoter() {}
+
+    /// @notice returns total balance of PCV in the Deposit
+    function balance() public view override returns (uint256) {
+        return _totalTokens(); // liquid and vote-escrowed tokens
```
Comment (Joeysantoro) [inline contracts/metagov/AngleDelegatorPCVDeposit.sol:37]: this method name is a bit confusing, perhaps _totalTokensManaged?

File: protocol-configuration/permissions.ts [hunk 21]
```diff
@@ -31,5 +30,9 @@ export const permissions = {
   PSM_ADMIN_ROLE: [],
   TRIBAL_CHIEF_ADMIN_ROLE: ['optimisticTimelock', 'tribalChiefSyncV2'],
   VOTIUM_ADMIN_ROLE: ['opsOptimisticTimelock'],
-  PCV_GUARDIAN_ADMIN_ROLE: ['optimisticTimelock']
+  PCV_GUARDIAN_ADMIN_ROLE: ['optimisticTimelock'],
+  METAGOVERNANCE_VOTE_ADMIN: ['feiDAOTimelock', 'opsOptimisticTimelock'],
+  METAGOVERNANCE_TOKEN_STAKING: ['feiDAOTimelock', 'opsOptimisticTimelock'],
+  METAGOVERNANCE_GAUGE_ADMIN: ['feiDAOTimelock', 'opsOptimisticTimelock'],
+  METAGOVERNANCE_GAUGE_STAKING: ['feiDAOTimelock', 'opsOptimisticTimelock']
```
Comment (Joeysantoro) [inline protocol-configuration/permissions.ts:37]: Are any of these higher clearance?

File: contracts/core/TribeRoles.sol (hunk 18 above)
Comment (eswak) [reply contracts/core/TribeRoles.sol:74]: Let's merge them, you're right it doesn't add much benefit

File: protocol-configuration/permissions.ts (hunk 21 above)
Comment (eswak) [reply protocol-configuration/permissions.ts:37]: arguably `METAGOVERNANCE_GAUGE_ADMIN` is higher risk because it can stake the protocol's LP tokens held on the contract on an arbitrary contract (it can whitelist a gauge & then stake in it), so I'll move it to the optimisticTimelock that has a longer duration

File: contracts/metagov/AngleDelegatorPCVDeposit.sol (hunk 20 above)
Comment (eswak) [reply contracts/metagov/AngleDelegatorPCVDeposit.sol:37]: changed to `_totalTokensManaged`

File: contracts/metagov/AngleDelegatorPCVDeposit.sol (hunk 19 above)
Comment (eswak) [reply contracts/metagov/AngleDelegatorPCVDeposit.sol:11]: Discussed in private, we'll keep it just in case Angle restore some kind of on-chain governance in the next 4 years

File: contracts/core/TribeRoles.sol (hunk 18 above)
Comment (eswak) [reply contracts/core/TribeRoles.sol:74]: `METAGOVERNANCE_GAUGE_STAKING` has been removed, and `METAGOVERNANCE_GAUGE_ADMIN` is used for everything gauge-related now

Comment (Joeysantoro) [inline contracts/core/TribeRoles.sol:74]: this role can be deleted

File: contracts/core/TribeRoles.sol [hunk 22]
```diff
@@ -58,6 +58,21 @@ library TribeRoles {
...
     bytes32 internal constant OPTIMISTIC_ADMIN = keccak256("OPTIMISTIC_ADMIN");
 
+    /// @notice manages meta-governance actions, like voting & delegating.
+    /// Also used to vote for gauge weights & similar liquid governance things.
+    bytes32 internal constant METAGOVERNANCE_VOTE_ADMIN = keccak256("METAGOVERNANCE_VOTE_ADMIN");
+
+    /// @notice allows to manage locking of vote-escrowed tokens, and staking/unstaking
+    /// governance tokens from a pre-defined contract in order to eventually allow voting.
+    /// Examples: ANGLE <> veANGLE, AAVE <> stkAAVE, CVX <> vlCVX, CRV > cvxCRV.
+    bytes32 internal constant METAGOVERNANCE_TOKEN_STAKING = keccak256("METAGOVERNANCE_TOKEN_STAKING");
```
Comment (Joeysantoro) [inline contracts/core/TribeRoles.sol:68]: I think METAGOVERNANCE_TOKEN_STAKING can also go under METAGOVERNANCE_GAUGE_ADMIN right?
